ITE/4/README.adoc

```diff
@@ -0,0 +1,598 @@
+`DISALLOW`, `CREATE`, `DELETE`, `REQUIRE`, and `MODIFY`. These rules perform
+different checks by verifying if artifacts matching the pattern occur or do not
+occur in the materials or products sections of the relevant link. This does not
+involve resolving the pattern or URI into the artifact itself, and so this ITE
+does not affect the working of these rules.
+
+However, it is quite likely that the contents referred to using generic URIs
+change more frequently than the traditional artifacts in the filesystem. The
+expectation is that either these changes are recorded with link attestations of
+their own, making them part of the supply chain, or metadata is replaced by
+authorized functionaries. The security implications of frequently changing
+generic resources are discussed in the Security section below.
+
+==== General Structure for URI Resolvers
+
+While the implementations of resolvers will vary wildly depending on the context
+they're written for, this ITE recommends they follow a certain structure that
+enables them to be reused by different stakeholders looking to hash the same
+abstract entities.
+
+```
+define hash_artifacts(string generic_uri, list hash_algorithms, kwargs) \
+    returns map artifact_hash_map
+
+define resolve_uri(string generic_uri, kwargs) returns list contained_uris
+
+define get_hashable_representation(string generic_uri, kwargs) returns bytes \
+    hashable_representation
+```
+
+`hash_artifacts` closely mirrors the general workflow of hashing artifacts in
+the filesystem, and is similar to how the in-toto reference implementations
+currently work. It requires generic artifacts to have a hashable representation
+of themselves. The method must return a map that contains all artifacts and
+their corresponding hash objects.
+
+`get_hashable_representation` is the fundamental component of any implementation
+that complies with this ITE. It is responsible for generating consistent
+representations of abstract entities while simultaneously maintaining their core
+properties. For example, an implementation that is capable of hashing GitHub
+entities may choose to use a subsection of the contents obtained using the
+official API, which returns JSON representations containing different properties
+of the entities. It is important for these representations to be consistent for
+a particular entity and they must be generated keeping in mind the different
+sources of non-determinism as highlighted in the Appendix. It's also necessary
+for this method to handle missing resources safely.
+
+Being able to get some hashable representation of abstract entities is necessary
+for in-toto inspections to verify contents as the context demands it. Therefore,
+it should be possible for implementations to use the method in scenarios other
+than just the recording of hashes of an artifact.
+
+It is quite likely that a generic URI specified may point to a collection of
```

Comment:
Could you explain how these functions relate with each other? For example, who is supposed to call whom? Maybe a small toy example would help.

Reply:
Do you think 8fc095b clarifies enough?